lib/ExtUtils/MM_Unix.pm

```diff
@@ -2077,6 +2076,11 @@ sub init_PERL {
     # already escaped spaces.
     $self->{FULLPERL} =~ tr/"//d if $Is{VMS};
 
+    # `dmake` can fail for image (aka, executable) names which start with double-quotes
+    # * push quote inward by at least one character (or the drive prefix, if present)
+    # * including any initial directory separator preserves the `file_name_is_absolute` property
+    $self->{FULLPERL} =~ s/^"(\S(:\\|:)?)/$1"/ if $self->is_make_type('dmake');
```

Comment:
Is there a reason not to use `Text::ParseWords` to unwrap, then some more conceptually correct thing, rather than just an RE?

Reply:
I hadn't considered that option.

I see your point that my solution here is just a mechanical work-around for the `dmake` bug. But, to me, conceptually, the FULLPERL string is correct, it's just that `dmake` can't handle an initial double-quote. So, a simple mechanical fix seems reasonable to me.

And, just quickly mentally spit-balling a variant solution using `Text::ParseWords`, I don't think that it would be an improvement. To my mind, splitting the string and reconstructing it into a `dmake`-palatable version (ie, sub-quoting pieces of the string) would be more error-prone.

I'm happy to be persuaded otherwise if you have a better construction in mind that you could share.

Reply:
I think that a parsewords, then a `File::Spec::splitdir`, then quoting the directory parts and filename using `MM->quote_literal`, would be the most reliable. REs have a way of going wrong.

Reply:
Hmm...

``` perl
$self->{FULLPERL} =~ s/^"(\S(:\\|:)?)/$1"/ if $self->is_make_type('dmake');
```

would be replaced with 

``` perl
$self->{FULLPERL} = do {
    use File::Spec q//; 
    use Text::ParseWords q//;
    my $p = $self->{FULLPERL}; 
    $p =~ s/\\/\\\\/gmsx;
    my @w = Text::ParseWords::shellwords($p);
    join(q/ /, File::Spec->catpath(map {MM->quote_literal($_)} File::Spec->splitpath(shift @w)), map {MM->quote_literal($_)} @w);
    } if $self->is_make_type('dmake');
```

or, if we can assume that `$self->{FULLPERL}` is a always a single "shellword", 

``` perl
 $self->{FULLPERL} = do {
    use File::Spec q//; 
    use Text::ParseWords q//;
    my $p = $self->{FULLPERL}; 
    $p =~ s/\\/\\\\/gmsx;
    File::Spec->catpath(map {MM->quote_literal($_)} File::Spec->splitpath((Text::ParseWords::shellwords($p))[0]));
    } if $self->is_make_type('dmake');
```

I can do something like that, but I'm hesitant. Do you think that's a better version?

Reply:
I have no experience to speak to what's more appropriate, but if you add use statements please do so at the top of the file or use require to lazy-load, it's misleading to put use statements in smaller scopes.

Reply:
A runtime `require` would also achieve the same as the `use...()`, as well as the lazy aspect mentioned above.

Reply:
Also I'm not sure of the purpose of the `\` related RE?

Reply:
This was just a first pass to show the steps that I thought were going to be needed.

@grinnz , I'll ultimately push the modules up in scope or use `require ...`.

@mohawk2 , paths in Win32 use backslashes which unfortunately is used as an escape character within the `Text::ParseWords` functions. I don't see a way to turn that off, so I pre-escaped the backslashes.

Am I getting that you both believe this is a better way to go?

Reply:
So, after testing and iterating on the code, it looks like this construction is what is required to replace the original RE (at both locations, with FULLPERL => PERL in the second location)...

``` perl
$self->{FULLPERL} = do {
    require Text::ParseWords;
    my $p = $self->{FULLPERL};
    $p =~ s/\\/\\\\/gmsx;
    my @w = Text::ParseWords::shellwords($p);
    if (!@w) { $p =~ s/^"//msx; @w = Text::ParseWords::shellwords($p); }
    join(q/ /, File::Spec->catpath(map {$self->quote_literal($_)} File::Spec->splitpath(shift @w)), map {$self->quote_literal($_)} @w);
    } if $self->is_make_type('dmake');
```

`$p =~ s/\\/\\\\/gmsx;` is required because `shellwords()` otherwise removes the backslashes from the path (commonly used for windows paths). And the ` if (!@w) { $p =~ s/^"//msx; @w = Text::ParseWords::shellwords($p); }` construction is required because some supplied FULLPERL/PERL strings have unbalanced double-quotes, which causes `shellwords()` to return an empty list. And the entire returned `shellwords()` array must be considered as perl flags may be attached to the string as a suffix.

I don't know of an available alternative to `shellwords()` which ignores backslash escapes and is forgiving of unbalanced double-quotes. If such a function exists, the code could be boiled down to a more palatable ...

``` perl
$self->{FULLPERL} = do {
    require ...;
    my @w = ALTERNATE_SHELLWORDS($self->{FULLPERL});
    join(q/ /, File::Spec->catpath(map {$self->quote_literal($_)} File::Spec->splitpath(shift @w)), map {$self->quote_literal($_)} @w);
    } if $self->is_make_type('dmake');
```

But, otherwise, I'm inclined to stick with the current RE. It's succinct and works in all tested `dmake` make-style perls.

I could limit it to just the MSWin32 `dmake` make-style perls, which are all tested via AppVeyor. That could limit any breakage on other perl variants (if any other `dmake` make-style variants actually exist), eg ...

``` perl
$self->{FULLPERL} =~ s/^"(\S(:\\|:)?)/$1"/ if ($self->is_make_type('dmake') && ($^O eq 'MSWin32'));
```

Reply:
As the misguided person who made EUMM work with gmake on Windows, and with spaces in paths, I am somewhat familiar with its quirks. I'd forgotten that ParseWords is effectively recreating the Bourne (etc) shell, rather than being per-platform. That being so, the `\` stuff is quite reasonable. I would encourage you to consider making a Win32 equivalent of the full function (see `Win32::ParseWords`) to be as bulletproof as possible, rather than doing "one bit at a time", which is an easy rabbithole to get sucked into.

Anyway, the ultimate goal is to make something that actually works - being minimally ugly is above all a means towards that, rather than an end to itself. Keep up the good work.

Reply:
Thanks for the feedback and conversation.

I'm going to leave this as is and resolve it, if that's ok with you.

Or, I can add the OS restriction (`$^O eq 'MSWin32'`) if you feel that narrowing the scope adds value. The OS restricted version still fixes the `dmake` bug and passes all tests.

I'll add creating a more portable `...::ParseWords` to my "ToDO list". Or, I might create a patch for `Text::ParseWords`, adding optional arguments and enabling more flexible use.

Reply:
If the limitation being dealt with is in dmake, keep it specific to that. There's no reason for it to apply to Win32 with gmake, I gather. By the way, a future version (not to feature-creep this version) will want to also test against gmake. That only takes an edit of `Config_heavy.pl`, from memory, but let's first get something that operates!

Reply:
@mohawk2 , FYI, the more recent versions of Strawberry Perl are using `gmake`. So, the current AppVeyor testing does test against both `dmake` and `gmake`.